### PR TITLE
qemu: Refactor `Qemu::run` method

### PR DESCRIPTION
**Pull request by chantra**
That method was made of multiple steps (boot vm, setup vm...) but it was all done within the same method.

This change creates to new method: boot_vm and setup_vm, taking responsibility of those individual steps.

This will likely help with #54

### CODE REVIEW

File: src/qemu.rs [hunk 1]
```diff
@@ -935,12 +938,44 @@ impl Qemu {
...
+        // stream.
+        let qmp_stream = qapi::Stream::new(
+            BufReader::new(
+                qmp_unix_stream
+                    .try_clone()
+                    .context("Failed to clone QMP Unix Stream")?,
+            ),
+            qmp_unix_stream,
+        );
+        let mut qmp = Qmp::new(qmp_stream);
```
Comment (danobi) [inline src/qemu.rs:968]: This is b/c we can't pass a reference to the `UnixStream` like we did before cuz now `qmp` outlives the `UnixStream`, right?

Comment (chantra) [reply src/qemu.rs:968]: correct.

File: src/qemu.rs [hunk 2]
```diff
@@ -921,39 +921,63 @@ impl Qemu {
...
+        let qga = QgaWrapper::new(self.qga_sock.clone(), host_supports_kvm(&self.arch));
+        let qga = match qga {
+            Ok(q) => q,
+            Err(e) => {
+                return Err(e).context("Failed to connect QGA");
+            }
+        };
+        let _ = self.updates.send(Output::BootEnd(Ok(())));
+
+        Ok((scopeguard::ScopeGuard::into_inner(child), qga))
```
[the remark this answers is not in the document]
Comment (chantra) [reply src/qemu.rs]: was fighting with:
```
error[E0308]: mismatched types
   --> src/qemu.rs:995:13
    |
995 |         Ok((child, qga, qmp))
    |             ^^^^^ expected `ScopeGuard<Child, fn(Child)>`, found `ScopeGuard<Child, fn(Child) {Qemu::child_cleanup}>`
    |
    = note: expected struct `ScopeGuard<_, fn(Child)>`
               found struct `ScopeGuard<_, fn(Child) {Qemu::child_cleanup}>`
```
happy to look back into it.

**Comment by chantra** [comment]
ok, the scopeguard thing is fixed.

File: src/qemu.rs [hunk 3]
```diff
@@ -939,12 +943,44 @@ impl Qemu {
...
+        let qmp_unix_stream = match connect_to_uds(&self.qmp_sock) {
+            Ok(s) => s,
+            Err(e) => {
+                let err = Self::extract_child_stderr(&mut child);
+                return Err(e).context("Failed to connect QMP").context(err);
+            }
+        };
+
+        // UnixStream is not directly cloneable, so we cannot use `Qmp::from_stream`
+        // without borrowoing the stream. Instead we create a new stream from the
```
Comment (danobi) [inline src/qemu.rs:962]: nit: borrowoing > borrowing

**Review by danobi — APPROVED** [review]
you have write access now on this repo, right?

**Comment by chantra** [comment]
> you have write access now on this repo, right?

Yes I do, but your feedback is appreciated  😄

**Comment by danobi** [comment]
Ha, I meant more that we can avoid an extra round trip for final comments. But yeah, for non-trivial stuff I am happy to review

**Pull request merged**